botcore/utils/cooldown.py

```diff
@@ -0,0 +1,184 @@
+
+    async def call_without_cooldown(self) -> R:
+        """
+        Run the command this cooldown blocked.
+
+        Returns:
+            The command's return value.
+        """
+        return await self._function(*self._args, **self._kwargs)
+
+
+@dataclass
+class _CooldownItem:
+    call_arguments: _ArgsTuple
+    timeout_timestamp: float
+
+
+class _CommandCooldownManager:
+    """
+    Manage invocation cooldowns for a command through the arguments the command is called with.
+
+    A cooldown is set through `set_cooldown` for a channel with the given `call_arguments`,
+    if `is_on_cooldown` is checked within `cooldown_duration` seconds
+    of the call to `set_cooldown` with the same arguments, True is returned.
+    """
+
+    def __init__(self, *, cooldown_duration: float):
+        self._cooldowns = dict[tuple[Hashable, _ArgsTuple], float]()
+        self._cooldowns_non_hashable = dict[Hashable, list[_CooldownItem]]()
+        self._cooldown_duration = cooldown_duration
+        self.cleanup_task = scheduling.create_task(
+            self._periodical_cleanup(random.uniform(0, 10)),
+            name="CooldownManager cleanup",
+        )
+
+    def set_cooldown(self, channel: Hashable, call_arguments: _ArgsTuple) -> None:
+        """Set `call_arguments` arguments on cooldown in `channel`."""
+        timeout_timestamp = time.monotonic() + self._cooldown_duration
+
+        try:
+            self._cooldowns[(channel, call_arguments)] = timeout_timestamp
+        except TypeError:
+            cooldowns_list = self._cooldowns_non_hashable.setdefault(channel, [])
+            for item in cooldowns_list:
+                if item.call_arguments == call_arguments:
+                    item.timeout_timestamp = timeout_timestamp
+            else:
+                cooldowns_list.append(_CooldownItem(call_arguments, timeout_timestamp))
+
+    def is_on_cooldown(self, channel: Hashable, call_arguments: _ArgsTuple) -> bool:
+        """Check whether ``call_arguments`` is on cooldown in ``channel``."""
+        current_time = time.monotonic()
+        try:
+            return self._cooldowns.get((channel, call_arguments), -math.inf) > current_time
+        except TypeError:
+            cooldowns_list = self._cooldowns_non_hashable.get(channel, None)
+            if cooldowns_list is None:
+                return False
+
+            for item in cooldowns_list:
+                if item.call_arguments == call_arguments:
+                    return item.timeout_timestamp > current_time
+            return False
+
+    async def _periodical_cleanup(self, initial_delay: float) -> None:
+        """
+        Wait for `initial_delay`, after that delete stale items every hour.
+
+        The `initial_delay` ensures we're not running cleanups for every command at the same time.
+        """
+        await asyncio.sleep(initial_delay)
+        while True:
+            await asyncio.sleep(60 * 60)
+            self._delete_stale_items()
```

Comment:
Why opt for a polling approach over a scheduling approach?

Reply:
I don't think a scheduler for each command would have much overhead if that's what you mean; it only needs a single dictionary to track state. Regardless, I don't have a problem with the polling approach if that's what you want to go with.